File: BDFM-RAG-System/src/services/rag.service.ts
import embeddingService from './embedding.service';
import qdrantService from './qdrant.service';
import llmService from './llm.service';
import logger from '../utils/logger';
import config from '../config';
import { SearchRequest, RAGResponse, SearchResult } from '../models';
import { extractHighlights } from '../utils/helpers';

export class RAGService {
  /**
   * Query using RAG
   */
  async query(request: SearchRequest): Promise<RAGResponse> {
    const startTime = Date.now();
    let embeddingTime = 0;
    let searchTime = 0;
    let generationTime = 0;

    try {
      logger.info(`Processing RAG query: "${request.query}"`);

      // 1. Generate embedding for query
      const embeddingStart = Date.now();
      const queryEmbedding = await embeddingService.generateEmbedding(
        request.query
      );
      embeddingTime = Date.now() - embeddingStart;
      logger.debug(`Embedding generated in ${embeddingTime}ms`);

      // 2. Search similar documents in Qdrant
      const searchStart = Date.now();
      const similarDocs = await qdrantService.searchSimilar(
        config.collections.correspondence,
        queryEmbedding,
        request.maxResults || config.rag.maxResults,
        request.similarityThreshold || config.rag.similarityThreshold,
        request.filters
      );
      searchTime = Date.now() - searchStart;
      logger.debug(
        `Search completed in ${searchTime}ms, found ${similarDocs.length} results`
      );

      // 3. Check if we found any results
      if (similarDocs.length === 0) {
        const language = request.language || 'ar';
        return {
          answer:
            language === 'ar'
              ? 'لم أجد أي مراسلات مشابهة لاستفسارك. يرجى المحاولة بكلمات مختلفة أو أكثر تحديداً.'
              : 'I couldn\'t find any correspondence similar to your query. Please try different or more specific terms.',
          sources: [],
          language,
          metadata: {
            queryProcessingTime: Date.now() - startTime,
            embeddingTime,
            searchTime,
            generationTime: 0,
          },
        };
      }

      // 4. Add highlights to results
      const resultsWithHighlights = this.addHighlights(
        similarDocs,
        request.query
      );

      // 5. Generate answer using LLM
      const generationStart = Date.now();
      const answer = await llmService.generateAnswer(
        request.query,
        resultsWithHighlights,
        request.language || 'ar'
      );
      generationTime = Date.now() - generationStart;
      logger.debug(`Answer generated in ${generationTime}ms`);

      const totalTime = Date.now() - startTime;
      logger.info(
        `RAG query completed in ${totalTime}ms (embedding: ${embeddingTime}ms, search: ${searchTime}ms, generation: ${generationTime}ms)`
      );

      return {
        answer,
        sources: resultsWithHighlights,
        language: request.language || 'ar',
        metadata: {
          queryProcessingTime: totalTime,
          embeddingTime,
          searchTime,
          generationTime,
        },
      };
    } catch (error: any) {
      logger.error('Error processing RAG query:', error);
      throw new Error(`Failed to process query: ${error.message}`);
    }
  }

  /**
   * Query with streaming response
   */
  async *queryStream(request: SearchRequest): AsyncGenerator<any, void, unknown> {
    try {
      logger.info(`Processing streaming RAG query: "${request.query}"`);

      // 1. Generate embedding for query
      const queryEmbedding = await embeddingService.generateEmbedding(
        request.query
      );

      // 2. Search similar documents
      const similarDocs = await qdrantService.searchSimilar(
        config.collections.correspondence,
        queryEmbedding,
        request.maxResults || config.rag.maxResults,
        request.similarityThreshold || config.rag.similarityThreshold,
        request.filters
      );

      // 3. Check if we found any results
      if (similarDocs.length === 0) {
        const language = request.language || 'ar';
        yield {
          type: 'answer',
          content:
            language === 'ar'
              ? 'لم أجد أي مراسلات مشابهة لاستفسارك.'
              : 'I couldn\'t find any correspondence similar to your query.',
        };
        yield {
          type: 'sources',
          content: [],
        };
        return;
      }

      // 4. Add highlights
      const resultsWithHighlights = this.addHighlights(
        similarDocs,
        request.query
      );

      // 5. Send sources first
      yield {
        type: 'sources',
        content: resultsWithHighlights,
      };

      // 6. Stream answer
      yield {
        type: 'answer_start',
        content: '',
      };

      for await (const chunk of llmService.generateAnswerStream(
        request.query,
        resultsWithHighlights,
        request.language || 'ar'
      )) {
        yield {
          type: 'answer_chunk',
          content: chunk,
        };
      }

      yield {
        type: 'answer_end',
        content: '',
      };
    } catch (error: any) {
      logger.error('Error processing streaming RAG query:', error);
      yield {
        type: 'error',
        content: error.message,
      };
    }
  }

  /**
   * Search similar correspondences without LLM generation
   */
  async search(
    query: string,
    maxResults?: number,
    threshold?: number,
    filters?: any
  ): Promise<SearchResult[]> {
    try {
      logger.info(`Searching for: "${query}"`);

      // Generate embedding
      const queryEmbedding = await embeddingService.generateEmbedding(query);

      // Search
      const results = await qdrantService.searchSimilar(
        config.collections.correspondence,
        queryEmbedding,
        maxResults || config.rag.maxResults,
        threshold || config.rag.similarityThreshold,
        filters
      );

      // Add highlights
      const resultsWithHighlights = this.addHighlights(results, query);

      logger.info(`Found ${resultsWithHighlights.length} similar results`);

      return resultsWithHighlights;
    } catch (error: any) {
      logger.error('Error searching:', error);
      throw new Error(`Failed to search: ${error.message}`);
    }
  }

  /**
   * Add highlights to search results
   */
  private addHighlights(
    results: SearchResult[],
    query: string
  ): SearchResult[] {
    return results.map((result) => ({
      ...result,
      highlights: extractHighlights(result.bodyText || result.subject, query),
    }));
  }

  /**
   * Get system status
   */
  async getStatus(): Promise<any> {
    try {
      // Check Qdrant
      const qdrantConnected = await qdrantService.checkConnection();
      const collections = await qdrantService.getCollections();
      const collectionCounts: Record<string, number> = {};

      for (const collection of collections) {
        collectionCounts[collection] =
          await qdrantService.getCollectionCount(collection);
      }

      // Check Ollama
      const ollamaConnected = await embeddingService.checkConnection();
      const models = await embeddingService.getAvailableModels();

      // Check PostgreSQL
      const postgresConnected = await this.checkPostgresConnection();

      return {
        qdrant: {
          connected: qdrantConnected,
          collections: collectionCounts,
        },
        ollama: {
          connected: ollamaConnected,
          models,
          embeddingModel: config.ollama.embeddingModel,
          chatModel: config.ollama.chatModel,
        },
        postgres: {
          connected: postgresConnected,
        },
        config: {
          embeddingDimension: config.rag.embeddingDimension,
          chunkSize: config.rag.chunkSize,
          maxResults: config.rag.maxResults,
          similarityThreshold: config.rag.similarityThreshold,
        },
      };
    } catch (error: any) {
      logger.error('Error getting status:', error);
      throw error;
    }
  }

  /**
   * Get all data from Qdrant collection
   */
  async getQdrantData(
    collectionName: string,
    limit: number = 100,
    offset: number = 0
  ): Promise<any> {
    try {
      logger.info(`Getting data from collection: ${collectionName}`);

      const qdrantService = (await import('./qdrant.service')).default;

      // Get collection info
      const collectionInfo = await qdrantService.getCollectionInfo(collectionName);

      // Get points from collection
      const points = await (qdrantService as any).client.scroll(collectionName, {
        limit,
        offset,
        with_payload: true,
        with_vector: false, // Don't include vectors to reduce response size
      });

      return {
        collection: collectionName,
        totalPoints: collectionInfo.points_count || 0,
        returnedPoints: points.points.length,
        limit,
        offset,
        data: points.points.map((point: any) => ({
          id: point.id,
          payload: point.payload,
        })),
      };
    } catch (error: any) {
      logger.error(`Error getting data from collection ${collectionName}:`, error);
      throw new Error(`Failed to get data from collection: ${error.message}`);
    }
  }

  /**
   * Get collection statistics
   */
  async getCollectionStats(collectionName: string): Promise<any> {
    try {
      logger.info(`Getting stats for collection: ${collectionName}`);

      const qdrantService = (await import('./qdrant.service')).default;

      const collectionInfo = await qdrantService.getCollectionInfo(collectionName);
      const count = await qdrantService.getCollectionCount(collectionName);

      return {
        collection: collectionName,
        pointsCount: count,
        vectorSize: collectionInfo.config?.params?.vectors?.size || 'unknown',
        distance: collectionInfo.config?.params?.vectors?.distance || 'unknown',
        status: collectionInfo.status || 'unknown',
        optimizerStatus: collectionInfo.optimizer_status || 'unknown',
        payloadSchema: collectionInfo.payload_schema || {},
        config: collectionInfo.config || {},
      };
    } catch (error: any) {
      logger.error(`Error getting stats for collection ${collectionName}:`, error);
      throw new Error(`Failed to get collection stats: ${error.message}`);
    }
  }

  /**
   * Get all collections
   */
  async getAllCollections(): Promise<any> {
    try {
      logger.info('Getting all Qdrant collections');

      const qdrantService = (await import('./qdrant.service')).default;

      const collections = await qdrantService.getCollections();

      // Get stats for each collection
      const collectionsWithStats = await Promise.all(
        collections.map(async (collectionName) => {
          try {
            const stats = await this.getCollectionStats(collectionName);
            return stats;
          } catch (error) {
            logger.warn(`Could not get stats for collection ${collectionName}:`, error);
            return {
              collection: collectionName,
              pointsCount: 0,
              vectorSize: 'unknown',
              distance: 'unknown',
              status: 'error',
              error: (error as Error).message,
            };
          }
        })
      );

      return {
        totalCollections: collections.length,
        collections: collectionsWithStats,
      };
    } catch (error: any) {
      logger.error('Error getting all collections:', error);
      throw new Error(`Failed to get collections: ${error.message}`);
    }
  }

  /**
   * Check PostgreSQL connection
   */
  private async checkPostgresConnection(): Promise<boolean> {
    try {
      const databaseService = (await import('./database.service')).default;
      return await databaseService.testConnection();
    } catch (error) {
      logger.error('Error checking PostgreSQL connection:', error);
      return false;
    }
  }
}

// Export singleton instance
export default new RAGService();

File: BDFM-RAG-System/bdfm-rag-python/app/services/rag_service.py
"""RAG service combining embedding, search, and LLM generation"""
from typing import List, Dict, Any, AsyncGenerator, Optional
import time
from app.config import settings
from app.utils.logger import logger
from app.utils.helpers import extract_highlights
from app.models import SearchRequest, QueryRequest, SearchResult, RAGResponse
from app.services.embedding_service import embedding_service
from app.services.qdrant_service import qdrant_service
from app.services.llm_service import llm_service


class RAGService:
    """Service for complete RAG pipeline"""
    
    async def query(self, request: QueryRequest) -> RAGResponse:
        """
        RAG query كامل: embedding + search + LLM generation
        
        Args:
            request: RAG query request
            
        Returns:
            الإجابة المولدة مع المصادر
        """
        start_time = time.time()
        embedding_time = 0
        search_time = 0
        generation_time = 0
        
        try:
            logger.info(f"Processing RAG query: {request.query[:50]}...")
            
            # 1. Generate embedding for query
            embedding_start = time.time()
            query_embedding = await embedding_service.generate_embedding(
                request.query,
                is_query=True  # مهم لـ E5 models
            )
            embedding_time = (time.time() - embedding_start) * 1000
            logger.debug(f"Embedding generated in {embedding_time:.0f}ms")
            
            # 2. Search similar documents in Qdrant
            search_start = time.time()
            similar_docs = await qdrant_service.search_similar(
                settings.correspondence_collection,
                query_embedding,
                limit=request.max_results or settings.max_results,
                score_threshold=request.similarity_threshold or settings.similarity_threshold,
                filters=request.filters
            )
            search_time = (time.time() - search_start) * 1000
            logger.debug(f"Search completed in {search_time:.0f}ms, found {len(similar_docs)} results")
            
            # 3. Check if we found any results
            if not similar_docs:
                language = request.language or "ar"
                no_results_msg = (
                    "لم أجد أي مراسلات مشابهة لاستفسارك. يرجى المحاولة بكلمات مختلفة أو أكثر تحديداً."
                    if language == "ar"
                    else "I couldn't find any correspondence similar to your query. Please try different or more specific terms."
                )
                
                return RAGResponse(
                    answer=no_results_msg,
                    sources=[],
                    language=language,
                    metadata={
                        "query_processing_time": (time.time() - start_time) * 1000,
                        "embedding_time": embedding_time,
                        "search_time": search_time,
                        "generation_time": 0
                    }
                )
            
            # 4. Add highlights to results
            results_with_highlights = self._add_highlights(similar_docs, request.query)
            
            # 5. Generate answer using LLM
            generation_start = time.time()
            answer = await llm_service.generate_answer(
                request.query,
                results_with_highlights,
                request.language or "ar"
            )
            generation_time = (time.time() - generation_start) * 1000
            logger.debug(f"Answer generated in {generation_time:.0f}ms")
            
            total_time = (time.time() - start_time) * 1000
            logger.info(
                f"RAG query completed in {total_time:.0f}ms "
                f"(embedding: {embedding_time:.0f}ms, search: {search_time:.0f}ms, generation: {generation_time:.0f}ms)"
            )
            
            return RAGResponse(
                answer=answer,
                sources=results_with_highlights,
                language=request.language or "ar",
                metadata={
                    "query_processing_time": total_time,
                    "embedding_time": embedding_time,
                    "search_time": search_time,
                    "generation_time": generation_time
                }
            )
            
        except Exception as e:
            logger.error(f"Error processing RAG query: {e}")
            raise
    
    async def query_stream(self, request: QueryRequest) -> AsyncGenerator[Dict[str, Any], None]:
        """
        RAG query مع streaming response
        
        Args:
            request: RAG query request
            
        Yields:
            أجزاء الإجابة والمصادر
        """
        try:
            logger.info(f"Processing streaming RAG query: {request.query[:50]}...")
            
            # 1. Generate embedding for query
            query_embedding = await embedding_service.generate_embedding(
                request.query,
                is_query=True
            )
            
            # 2. Search similar documents
            similar_docs = await qdrant_service.search_similar(
                settings.correspondence_collection,
                query_embedding,
                limit=request.max_results or settings.max_results,
                score_threshold=request.similarity_threshold or settings.similarity_threshold,
                filters=request.filters
            )
            
            # 3. Check if we found results
            if not similar_docs:
                language = request.language or "ar"
                yield {
                    "type": "answer",
                    "content": (
                        "لم أجد أي مراسلات مشابهة لاستفسارك."
                        if language == "ar"
                        else "I couldn't find any correspondence similar to your query."
                    )
                }
                yield {
                    "type": "sources",
                    "content": []
                }
                return
            
            # 4. Add highlights
            results_with_highlights = self._add_highlights(similar_docs, request.query)
            
            # 5. Send sources first
            yield {
                "type": "sources",
                "content": [result.dict() for result in results_with_highlights]
            }
            
            # 6. Stream answer
            yield {
                "type": "answer_start",
                "content": ""
            }
            
            async for chunk in llm_service.generate_answer_stream(
                request.query,
                results_with_highlights,
                request.language or "ar"
            ):
                yield {
                    "type": "answer_chunk",
                    "content": chunk
                }
            
            yield {
                "type": "answer_end",
                "content": ""
            }
            
        except Exception as e:
            logger.error(f"Error processing streaming RAG query: {e}")
            yield {
                "type": "error",
                "content": str(e)
            }
    
    async def search(
        self,
        query: str,
        max_results: Optional[int] = None,
        threshold: Optional[float] = None,
        filters: Optional[Dict[str, Any]] = None
    ) -> List[SearchResult]:
        """
        بحث دلالي بدون LLM generation
        
        Args:
            query: الاستعلام
            max_results: عدد النتائج
            threshold: عتبة التشابه
            filters: فلاتر إضافية
            
        Returns:
            نتائج البحث
        """
        try:
            logger.info(f"Searching for: {query[:50]}...")
            
            # Generate embedding
            query_embedding = await embedding_service.generate_embedding(query, is_query=True)
            
            # Search
            results = await qdrant_service.search_similar(
                settings.correspondence_collection,
                query_embedding,
                limit=max_results or settings.max_results,
                score_threshold=threshold or settings.similarity_threshold,
                filters=filters
            )
            
            # Add highlights
            results_with_highlights = self._add_highlights(results, query)
            
            logger.info(f"Found {len(results_with_highlights)} similar results")
            
            return results_with_highlights
            
        except Exception as e:
            logger.error(f"Error searching: {e}")
            raise
    
    def _add_highlights(
        self,
        results: List[Dict[str, Any]],
        query: str
    ) -> List[SearchResult]:
        """
        إضافة highlights لنتائج البحث
        
        Args:
            results: نتائج البحث
            query: الاستعلام
            
        Returns:
            نتائج مع highlights
        """
        results_with_highlights = []
        
        for result in results:
            highlights = extract_highlights(
                result.get("body_text", "") or result.get("subject", ""),
                query
            )
            
            search_result = SearchResult(
                id=result["id"],
                mail_num=result["mail_num"],
                mail_date=result["mail_date"],
                subject=result["subject"],
                body_text=result["body_text"],
                correspondence_type=result["correspondence_type"],
                secrecy_level=result["secrecy_level"],
                priority_level=result["priority_level"],
                personality_level=result["personality_level"],
                similarity_score=result["similarity_score"],
                highlights=highlights
            )
            
            results_with_highlights.append(search_result)
        
        return results_with_highlights
    
    async def get_status(self) -> Dict[str, Any]:
        """
        الحصول على حالة النظام
        
        Returns:
            معلومات الحالة
        """
        try:
            # Check Qdrant
            qdrant_connected = await qdrant_service.check_connection()
            collections = await qdrant_service.get_collections()
            
            collection_counts = {}
            for collection in collections:
                count = await qdrant_service.get_collection_count(collection)
                collection_counts[collection] = count
            
            # Check Ollama
            ollama_connected = await llm_service.check_connection()
            
            # Check Embedding model
            embedding_loaded = await embedding_service.check_model_loaded()
            
            return {
                "qdrant": {
                    "connected": qdrant_connected,
                    "collections": collection_counts
                },
                "ollama": {
                    "connected": ollama_connected,
                    "chat_model": settings.ollama_chat_model
                },
                "embedding": {
                    "model": settings.embedding_model,
                    "dimension": settings.embedding_dimension,
                    "loaded": embedding_loaded
                },
                "config": {
                    "chunk_size": settings.chunk_size,
                    "max_results": settings.max_results,
                    "similarity_threshold": settings.similarity_threshold
                }
            }
            
        except Exception as e:
            logger.error(f"Error getting status: {e}")
            raise


# Create global RAG service instance
rag_service = RAGService()


__all__ = ["rag_service", "RAGService"]

